### DISCUSSION TITLE
validateDatabaseSchema throws "Contains the following unexpected entried: android_metadata"

### DISCUSSION THREAD

**Discussion by madsane29** [simolus3/drift#2042] [Q&A] [answered]
I try to validate the schema of my database and throws that exception. I wonder what is the problem?

![image](https://user-images.githubusercontent.com/40585482/189478900-7683d26a-2bc4-4fef-a86a-51ee6a4d1826.png)

```
E/flutter (11274): [ERROR:flutter/lib/ui/ui_dart_state.cc(198)] Unhandled Exception: Schema does not match
E/flutter (11274): additional:
E/flutter (11274):  Contains the following unexpected entries: android_metadata
E/flutter (11274): 
E/flutter (11274): #0      verify (package:drift_dev/src/services/schema/verifier_impl.dart:132:5)
E/flutter (11274): #1      VerifySelf.validateDatabaseSchema (package:drift_dev/api/migrations.dart:96:5)
E/flutter (11274): <asynchronous suspension>
E/flutter (11274): #2      AppDatabase.migration.<anonymous closure> (package:inkmobile/repository/database.dart:90:13)
E/flutter (11274): <asynchronous suspension>
E/flutter (11274): #3      GeneratedDatabase.beforeOpen.<anonymous closure> (package:drift/src/runtime/api/db_base.dart:120:7)
E/flutter (11274): <asynchronous suspension>
E/flutter (11274): #4      DelegatedDatabase._runMigrations (package:drift/src/runtime/executor/helpers/engines.dart:368:5)
E/flutter (11274): <asynchronous suspension>
E/flutter (11274): #5      DelegatedDatabase.ensureOpen.<anonymous closure> (package:drift/src/runtime/executor/helpers/engines.dart:336:7)
E/flutter (11274): <asynchronous suspension>
E/flutter (11274): #6      InsertStatement.insert (package:drift/src/runtime/query_builder/statements/insert.dart:73:12)
E/flutter (11274): <asynchronous suspension>
```

Thanks in advance

**Comment by simolus3** [comment, marked as answer]
How do you open your database? Are you using `drift_sqflite` or a `NativeDatabase`?

It looks like that table was created automatically in older Android versions to implement the  settings provider. Now, the question is how it ended up in your database :) AFAIK, it's not created by default for each database. I could just ignore those tables, but `bluethooth_devices`, `bookmarks`, `favorites` and `old_favorites` are similar tables that may have been created by the setting manager and I don't want to ignore all of them since they could very well be the name of tables that drift users as creating as well. So I'd like to understand how those tables ended up existing in your database.

**Reply by madsane29** [reply]
I did not migrate to drift yet... It is an older project, I have these dependencies:

```
dependencies:
  moor: ^4.3.2
  moor_flutter: ^4.0.0
  moor_db_viewer: ^4.3.0

dev_dependencies:
   moor_generator: ^4.3.1
```

![image](https://user-images.githubusercontent.com/40585482/189578055-d8b7f3ea-432f-4b84-8c8b-ae6273a79176.png)


Then I guess I ignore this error until the migration happens... Thanks.

**Reply by simolus3** [reply]
This shouldn't happen with `moor_flutter` either (which is pretty much the same as `drift_sqflite`). Are you exclusively using that database in Dart or did you ever use it in native Android code?

Can you run `SELECT * FROM sqlite_master` in a custom statement and see if there are any other tables you didn't expect? If this turns out to be a thing that Android just does I will change `drift_dev` to ignore the `android_metadata` table.

**Reply by madsane29** [reply]
Never touched android in this project. 

This exception is thrown even if I delete the app from the device (I use a physical device for testing, Xiaomi Redmi Note 8 Pro) and install it again through android studio. Also checked on emulator and same result (API 31).

These are the 2 tables that I found and I did not create it (used `'SELECT * FROM sqlite_master;'`):

```
I/flutter (15889): {type: table, name: android_metadata, tbl_name: android_metadata, rootpage: 3, sql: CREATE TABLE android_metadata (locale TEXT)}
I/flutter (15889): {type: table, name: sqlite_sequence, tbl_name: sqlite_sequence, rootpage: 5, sql: CREATE TABLE sqlite_sequence(name,seq)}
```

**Reply by simolus3** [reply]
Alright, I'll just ignore that table then: 4da647e8a3f90acdb53dba2108ff0cfe38eadc53
